File: client/src/lib/games/isolation/voronoi.ts
/**
 * Voronoi Territory Analysis for ISOLATION
 *
 * Calculates which cells each player can reach first using BFS.
 * Now uses bitboard operations for improved performance.
 */

import type { BoardState } from "./types";
import {
  calculateBitboardVoronoi,
  posToIndex,
  indexToPos,
  CELL_MASKS,
  getQueenMoves,
  bitboardToIndices,
  popCount
} from "./bitboard";

// 8 directions for queen movement
const DIRECTIONS = [
  { dr: -1, dc: -1 }, { dr: -1, dc: 0 }, { dr: -1, dc: 1 },
  { dr: 0, dc: -1 },                      { dr: 0, dc: 1 },
  { dr: 1, dc: -1 },  { dr: 1, dc: 0 },  { dr: 1, dc: 1 }
];

export interface VoronoiResult {
  playerTerritory: number;   // Cells player reaches first
  aiTerritory: number;       // Cells AI reaches first
  contested: number;         // Cells reached at same distance
  playerDistance: number[];  // Distance from player to each cell (49 cells)
  aiDistance: number[];      // Distance from AI to each cell (49 cells)
  totalReachable: number;    // Total reachable cells by either player
}

/**
 * Calculate Voronoi territories for both players
 * Uses bitboard-based BFS for performance
 */
export function calculateVoronoi(board: BoardState): VoronoiResult {
  const { boardSize, playerPos, aiPos, destroyed } = board;
  const BOARD_SIZE = boardSize.rows * boardSize.cols;
  const INF = 999;

  // Use bitboard implementation for territory calculation
  const bbVoronoi = calculateBitboardVoronoi(playerPos, aiPos, destroyed);

  // Calculate distances using BFS for compatibility
  const playerDist = new Array(BOARD_SIZE).fill(INF);
  const aiDist = new Array(BOARD_SIZE).fill(INF);

  // BFS from both positions
  bfsDistance(playerPos, playerDist, aiPos, boardSize, destroyed);
  bfsDistance(aiPos, aiDist, playerPos, boardSize, destroyed);

  return {
    playerTerritory: bbVoronoi.playerCount,
    aiTerritory: bbVoronoi.aiCount,
    contested: bbVoronoi.contestedCount,
    playerDistance: playerDist,
    aiDistance: aiDist,
    totalReachable: bbVoronoi.playerCount + bbVoronoi.aiCount + bbVoronoi.contestedCount
  };
}

/**
 * BFS to calculate shortest distance from start position to all cells
 * Each queen move counts as distance 1 regardless of distance traveled
 */
function bfsDistance(
  startPos: { r: number; c: number },
  distances: number[],
  otherPiecePos: { r: number; c: number },
  boardSize: { rows: number; cols: number },
  destroyed: { r: number; c: number }[]
): void {
  const INF = 999;

  const isBlocked = (r: number, c: number): boolean => {
    if (r < 0 || r >= boardSize.rows || c < 0 || c >= boardSize.cols) return true;
    if (r === otherPiecePos.r && c === otherPiecePos.c) return true;
    return destroyed.some(d => d.r === r && d.c === c);
  };

  distances.fill(INF);
  const startIdx = posToIndex(startPos.r, startPos.c);
  distances[startIdx] = 0;

  const queue: Array<{ r: number; c: number; dist: number }> = [
    { r: startPos.r, c: startPos.c, dist: 0 }
  ];

  while (queue.length > 0) {
    const { r, c, dist } = queue.shift()!;

    for (const dir of DIRECTIONS) {
      let nr = r + dir.dr;
      let nc = c + dir.dc;

      while (!isBlocked(nr, nc)) {
        const nextIdx = posToIndex(nr, nc);

        if (distances[nextIdx] > dist + 1) {
          distances[nextIdx] = dist + 1;
          queue.push({ r: nr, c: nc, dist: dist + 1 });
        }

        nr += dir.dr;
        nc += dir.dc;
      }
    }
  }
}

/**
 * Quick mobility calculation - number of immediate valid moves
 */
export function calculateImmediateMobility(
  pos: { r: number; c: number },
  otherPos: { r: number; c: number },
  boardSize: { rows: number; cols: number },
  destroyed: { r: number; c: number }[]
): number {
  // Create blocked bitboard
  let blocked = 0n;
  for (const d of destroyed) {
    blocked |= CELL_MASKS[posToIndex(d.r, d.c)];
  }
  blocked |= CELL_MASKS[posToIndex(otherPos.r, otherPos.c)];

  const moves = getQueenMoves(pos, blocked);
  return popCount(moves);
}

/**
 * Calculate the "critical cells" - cells that divide the board
 */
export function identifyCriticalCells(board: BoardState): number[] {
  const { boardSize, playerPos, aiPos, destroyed } = board;
  const critical: number[] = [];

  const voronoi = calculateVoronoi(board);

  for (let r = 0; r < boardSize.rows; r++) {
    for (let c = 0; c < boardSize.cols; c++) {
      const idx = posToIndex(r, c);

      // Skip destroyed and occupied cells
      if (destroyed.some(d => d.r === r && d.c === c)) continue;
      if ((r === playerPos.r && c === playerPos.c) ||
          (r === aiPos.r && c === aiPos.c)) continue;

      const pDist = voronoi.playerDistance[idx];
      const aDist = voronoi.aiDistance[idx];

      if (pDist === 999 && aDist === 999) continue;

      // Critical if contested or on the frontier
      if (Math.abs(pDist - aDist) <= 1) {
        critical.push(idx);
      }
    }
  }

  return critical;
}

File: client/src/lib/games/isolation/wasm/src/voronoi.rs
//! Voronoi Territory Analysis for ISOLATION
//!
//! Calculates which cells each player can reach first using bitboard-based BFS.
//! This is a critical component of strategic evaluation (weight 5.0).
//!
//! Based on TypeScript implementation in bitboard.ts:calculateBitboardVoronoiOptimized()

use crate::board::GameState;
use crate::bitboard::*;

#[derive(Clone, Copy, Debug)]
pub struct VoronoiResult {
    pub player_territory: u64,
    pub ai_territory: u64,
    pub contested: u64,
    pub player_count: i32,
    pub ai_count: i32,
    pub contested_count: i32,
}

/// Calculate Voronoi territories using optimized bitboard-based BFS
///
/// This is ~50-70% faster than distance-based approaches because:
/// - No distance array allocation (eliminates 98 number allocations)
/// - Direct bit manipulation (no array indexing)
/// - Territory determined by frontier reach order, not distance comparison
///
/// Returns territory bitboards and counts for both players
pub fn calculate_voronoi_optimized(
    player_pos: (u8, u8),
    ai_pos: (u8, u8),
    destroyed: u64,
) -> VoronoiResult {
    // Create blocked bitboard (destroyed + both positions)
    let player_idx = pos_to_index(player_pos.0, player_pos.1);
    let ai_idx = pos_to_index(ai_pos.0, ai_pos.1);

    let player_mask = 1u64 << player_idx;
    let ai_mask = 1u64 << ai_idx;
    let blocked = destroyed | player_mask | ai_mask;

    // Initialize frontiers
    let mut player_frontier = player_mask;
    let mut ai_frontier = ai_mask;
    let mut player_visited = player_mask;
    let mut ai_visited = ai_mask;

    // Territory bitboards - accumulated as frontiers expand
    let mut player_territory = 0u64;
    let mut ai_territory = 0u64;
    let mut contested = 0u64;

    let max_depth = 20; // Safety limit
    let mut depth = 0;

    // Dual-frontier expansion WITHOUT distance arrays
    while (player_frontier != 0 || ai_frontier != 0) && depth < max_depth {
        depth += 1;

        // 1. Expand both frontiers simultaneously
        let new_player_frontier = if player_frontier != 0 {
            expand_frontier_optimized(player_frontier, blocked, player_visited)
        } else {
            0
        };

        let new_ai_frontier = if ai_frontier != 0 {
            expand_frontier_optimized(ai_frontier, blocked, ai_visited)
        } else {
            0
        };

        // 2. Determine ownership based on simultaneous arrival
        
        // Player claims: Reached by player, NOT visited by AI before, NOT reached by AI now
        let player_only = new_player_frontier & !ai_visited & !new_ai_frontier;
        
        // AI claims: Reached by AI, NOT visited by player before, NOT reached by player now
        let ai_only = new_ai_frontier & !player_visited & !new_player_frontier;

        // Contested: Reached by BOTH in this same step
        let contested_new = new_player_frontier & new_ai_frontier;

        // 3. Update territory accumulators
        player_territory |= player_only;
        ai_territory |= ai_only;
        contested |= contested_new;

        // 4. Update visited sets
        player_visited |= new_player_frontier;
        ai_visited |= new_ai_frontier;

        // 5. Advance frontiers
        player_frontier = new_player_frontier;
        ai_frontier = new_ai_frontier;
    }

    VoronoiResult {
        player_territory,
        ai_territory,
        contested,
        player_count: count_ones(player_territory) as i32,
        ai_count: count_ones(ai_territory) as i32,
        contested_count: count_ones(contested) as i32,
    }
}

/// Expand frontier by one queen-move step (bit-parallel optimization)
///
/// Returns only new cells (not already visited).
fn expand_frontier_optimized(
    frontier: u64,
    blocked: u64,
    visited: u64,
) -> u64 {
    expand_queen_bit_parallel(frontier, blocked) & !visited
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_voronoi_basic() {
        // Simple case: player at (0,0), AI at (6,6), no destroyed cells
        let player_pos = (0, 0);
        let ai_pos = (6, 6);
        let destroyed = 0u64;

        let result = calculate_voronoi_optimized(player_pos, ai_pos, destroyed);

        // Player should control top-left area, AI should control bottom-right
        assert!(result.player_count > 0);
        assert!(result.ai_count > 0);

        // Total reachable should be less than full board (49) minus 2 positions
        let total = result.player_count + result.ai_count + result.contested_count;
        assert!(total <= 47); // 49 - 2 occupied cells
    }

    #[test]
    fn test_voronoi_contested() {
        // Players equidistant from center: player at (3,0), AI at (3,6)
        let player_pos = (3, 0);
        let ai_pos = (3, 6);
        let destroyed = 0u64;

        let result = calculate_voronoi_optimized(player_pos, ai_pos, destroyed);

        // Should have significant contested territory in the middle
        assert!(result.contested_count > 0);
    }

    #[test]
    fn test_voronoi_blocked() {
        // Test with some destroyed cells blocking paths
        let player_pos = (0, 0);
        let ai_pos = (6, 6);

        // Create a diagonal wall
        let mut destroyed = 0u64;
        for i in 0..7 {
            destroyed |= 1u64 << pos_to_index(i, i);
        }
        // Remove player and AI positions from wall
        destroyed &= !(1u64 << pos_to_index(0, 0));
        destroyed &= !(1u64 << pos_to_index(6, 6));

        let result = calculate_voronoi_optimized(player_pos, ai_pos, destroyed);

        // With diagonal wall, territories should be more separated
        assert!(result.player_count > 0);
        assert!(result.ai_count > 0);
    }
}
